Add admin CRUD views for managing organizations

robertknight opened the pull request:
Add views for:

 - Listing organizations
 - Creating and editing organizations
 - Deleting organizations

Since our form library does not have a file upload widget, the logo
input is just a text area where the user is expected to paste the SVG
content of the logo.

Organizations can only be deleted if they are not associated with any
groups.

Fixes https://github.com/hypothesis/product-backlog/issues/522

robertknight commented:
The `test_it_returns_filtered_orgs` test failed on Travis but I'm not able to reproduce locally. I'll look into this next week.

robertknight commented:
> The test_it_returns_filtered_orgs test failed on Travis but I'm not able to reproduce locally. I'll look into this next week.

This was happening because https://github.com/hypothesis/h/pull/4913 was merged into the DB since I forked my branch and Travis was testing a merge of this branch with master.

lyzadanger commented on h/views/admin_organizations.py:
```diff
@@ -0,0 +1,148 @@
+    @view_config(request_method='GET')
+    def get(self):
+        self.form.set_appstruct({
+            'authority': self.request.authority,
```
Thanks for pre-populating this!

lyzadanger commented on h/views/admin_organizations.py:
```diff
@@ -0,0 +1,148 @@
+
+            self.request.db.add(org)
+            self.request.session.flash(
+                Markup('Created new organization {}'.format(name)), 'success')
```
i18n here?

robertknight replied:
Good catch.

lyzadanger commented on h/views/admin_organizations.py:
```diff
@@ -0,0 +1,148 @@
+
+        def on_success(appstruct):
+            org.name = appstruct['name']
+            org.logo = appstruct['logo']
```
Naivete-with-SQLAlchemy alert! I was looking at the code here trying to spot the place where the db gets updated, but I presume that merely updating/dirtying the model object here will cause the changes to be persisted, correct?

robertknight replied:
Yes. Changes to model objects get committed at the end of the web request (see [my writeup of how this works](https://robertknight.me.uk/posts/pyramid-db-transactions/)) if they have not already been explicitly flushed.

The main reason for having explicit flushes in our code is to get the value of some field (eg. a new group's `pubid` or most objects' `id` fields) which is only generated when the object is saved.

lyzadanger commented on tests/h/schemas/admin_organization_test.py:
```diff
@@ -0,0 +1,27 @@
+
+
+class TestOrganizationSchema(object):
+    def test_it_allows_valid_data(self, org_data, bound_schema):
```
Is there any value for testing for name validation (too short, too long?)?

robertknight replied:
The minimum length is one char, so it can't be too short. It can be too long however. I'll add a test for this.

lyzadanger commented on tests/h/views/admin_organizations_test.py:
```diff
@@ -0,0 +1,175 @@
+
+        assert org.name == 'Updated name'
+        assert org.logo == '<svg>new logo</svg>'
+        assert ctx['form'] == self._expected_form(org)
```
Of course! These test classes are just classes after all; wrapping up stuff in methods is a completely reasonable way to do these sorts of things (it had never occurred to me before!).

lyzadanger commented on h/templates/admin/organizations_edit.html.jinja2:
```diff
@@ -0,0 +1,14 @@
+
+{% block content %}
+  {{ form }}
+  {% if delete_url %}
```
The positioning of this delete button is a bit wonky for me:

![image](https://user-images.githubusercontent.com/439947/38216577-352856f6-3699-11e8-8509-05e9c8da055f.png)

robertknight replied:
Yep. Zero points for beauty here. My plan was to get it in production and solicit some suggestions, hoping that we can fix the other views (OAuth clients, groups) at the same time.

robertknight commented on h/schemas/admin_organization.py:
```diff
@@ -0,0 +1,42 @@
+    logo = colander.SchemaNode(
+        colander.String(),
+        title=_('Logo'),
+        hint=_('Logo in SVG format'),
```
Yes. I'll add a note explaining in a bit more detail.

robertknight commented on h/resources.py:
```diff
@@ -114,7 +115,12 @@ def __init__(self, request):
+            org = self.request.db.query(Organization).filter_by(pubid=pubid).one()
+
+            # Inherit global ACL. See comments in `AuthClientFactory`.
+            org.__parent__ = Root(self.request)
```
Correct. Without this ACL inheritance, Pyramid will find the context object for the request (an `Organization`), look for a `__acl__` property on it, not find one, and then reject the view because the user "doesn't have the `admin_organizations`" permission.

robertknight commented:
I've addressed the comments except for the navbar and delete button. I agree that the navbar issue needs to be resolved. I'll work on that in a separate PR.
